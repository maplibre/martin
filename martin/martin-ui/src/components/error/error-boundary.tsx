'use client';

import { AlertTriangle, RefreshCw } from 'lucide-react';
import React from 'react';
import { Button } from '@/components/ui/button';
import { Card, CardContent, CardDescription, CardHeader, CardTitle } from '@/components/ui/card';

interface ErrorBoundaryState {
  hasError: boolean;
  error?: Error;
  errorInfo?: React.ErrorInfo;
}

interface ErrorBoundaryProps {
  children: React.ReactNode;
  fallback?: React.ComponentType<{ error?: Error; retry: () => void }>;
  onError?: (error: Error, errorInfo: React.ErrorInfo) => void;
}

class ErrorBoundary extends React.Component<ErrorBoundaryProps, ErrorBoundaryState> {
  constructor(props: ErrorBoundaryProps) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(error: Error): ErrorBoundaryState {
    return { error, hasError: true };
  }

  componentDidCatch(error: Error, errorInfo: React.ErrorInfo) {
    this.setState({ error, errorInfo });
    this.props.onError?.(error, errorInfo);
  }

  retry = () => {
    this.setState({ error: undefined, errorInfo: undefined, hasError: false });
  };

  render() {
    if (this.state.hasError) {
      const FallbackComponent = this.props.fallback || DefaultErrorFallback;
      return <FallbackComponent error={this.state.error} retry={this.retry} />;
    }

    return this.props.children;
  }
}

function DefaultErrorFallback({ error, retry }: { error?: Error; retry: () => void }) {
  return (
    <Card className="w-full max-w-md mx-auto mt-8">
      <CardHeader>
        <div className="flex items-center space-x-2">
          <AlertTriangle className="h-5 w-5 text-destructive" />
          <CardTitle className="text-destructive">Something went wrong</CardTitle>
        </div>
        <CardDescription>
          An unexpected error occurred. Please try again or contact support if the problem persists.
        </CardDescription>
      </CardHeader>
      <CardContent className="space-y-4">
        {error && (
          <div className="p-3 bg-destructive rounded-md">
            <p className="text-sm text-destructive-foreground font-mono">{error.message}</p>
          </div>
        )}
        <Button className="w-full" onClick={retry}>
          <RefreshCw className="h-4 w-4 mr-2" />
          Try Again
        </Button>
      </CardContent>
    </Card>
  );
}

export { ErrorBoundary };
