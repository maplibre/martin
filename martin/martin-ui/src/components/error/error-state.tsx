'use client';

import { AlertTriangle, RefreshCw, WifiOff } from 'lucide-react';
import { Button } from '@/components/ui/button';
import { Card, CardContent, CardDescription, CardHeader, CardTitle } from '@/components/ui/card';

export interface ErrorStateProps {
  title?: string;
  description?: string;
  error?: Error | string;
  onRetry?: () => void;
  isRetrying?: boolean;
  variant?: 'network' | 'server' | 'generic' | 'timeout';
  showDetails?: boolean;
}

export function ErrorState({
  title,
  description,
  error,
  onRetry,
  isRetrying = false,
  variant = 'generic',
  showDetails = false,
}: ErrorStateProps) {
  const getErrorConfig = () => {
    switch (variant) {
      case 'network':
        return {
          defaultDescription:
            'Unable to connect to the server. Please check your internet connection.',
          defaultTitle: 'Network Error',
          icon: <WifiOff className="h-8 w-8 text-destructive-foreground" />,
        };
      case 'server':
        return {
          defaultDescription: 'The server encountered an error. Please try again later.',
          defaultTitle: 'Server Error',
          icon: <AlertTriangle className="h-8 w-8 text-destructive-foreground" />,
        };
      case 'timeout':
        return {
          defaultDescription: 'The request took too long to complete. Please try again.',
          defaultTitle: 'Request Timeout',
          icon: <AlertTriangle className="h-8 w-8 text-orange-500" />,
        };
      default:
        return {
          defaultDescription: 'An unexpected error occurred. Please try again.',
          defaultTitle: 'Something went wrong',
          icon: <AlertTriangle className="h-8 w-8 text-destructive-foreground" />,
        };
    }
  };

  const config = getErrorConfig();
  const errorMessage = typeof error === 'string' ? error : error?.message;

  return (
    <Card className="w-full max-w-md mx-auto">
      <CardHeader className="text-center">
        <div className="flex justify-center mb-4">{config.icon}</div>
        <CardTitle className="text-lg">{title || config.defaultTitle}</CardTitle>
        <CardDescription>{description || config.defaultDescription}</CardDescription>
      </CardHeader>
      <CardContent className="space-y-4">
        {showDetails && errorMessage && (
          <div className="p-3 bg-destructive rounded-md border border-red-200">
            <p className="text-sm text-destructive-foreground font-mono break-words">
              {errorMessage}
            </p>
          </div>
        )}
        {onRetry && (
          <Button className="w-full" disabled={isRetrying} onClick={onRetry}>
            <RefreshCw className={`h-4 w-4 mr-2 ${isRetrying ? 'animate-spin' : ''}`} />
            {isRetrying ? 'Retrying...' : 'Try Again'}
          </Button>
        )}
      </CardContent>
    </Card>
  );
}

export function InlineErrorState({
  message,
  onRetry,
  isRetrying = false,
  variant = 'generic',
}: {
  message: string;
  onRetry?: () => void;
  isRetrying?: boolean;
  variant?: 'network' | 'server' | 'generic';
}) {
  const getIcon = () => {
    switch (variant) {
      case 'network':
        return <WifiOff className="h-4 w-4 text-destructive-foreground" />;
      case 'server':
        return <AlertTriangle className="h-4 w-4 text-destructive-foreground" />;
      default:
        return <AlertTriangle className="h-4 w-4 text-destructive-foreground" />;
    }
  };

  return (
    <div className="flex items-center justify-between p-3 bg-destructive border border-red-200 rounded-md">
      <div className="flex items-center space-x-2">
        {getIcon()}
        <span className="text-sm destructive-foreground">{message}</span>
      </div>
      {onRetry && (
        <Button disabled={isRetrying} onClick={onRetry} size="sm" variant="outline">
          <RefreshCw className={`h-3 w-3 mr-1 ${isRetrying ? 'animate-spin' : ''}`} />
          {isRetrying ? 'Retrying' : 'Retry'}
        </Button>
      )}
    </div>
  );
}
