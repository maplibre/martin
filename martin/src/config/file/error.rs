use std::path::PathBuf;

pub type ConfigFileResult<T> = Result<T, ConfigFileError>;

#[derive(thiserror::Error, Debug)]
pub enum ConfigFileError {
    #[error("IO error {0}: {1}")]
    IoError(#[source] std::io::Error, PathBuf),

    #[error("Unable to load config file {1}: {0}")]
    ConfigLoadError(#[source] std::io::Error, PathBuf),

    #[error("Unable to parse config file {1}: {0}")]
    ConfigParseError(#[source] subst::yaml::Error, PathBuf),

    #[error("Unable to write config file {1}: {0}")]
    ConfigWriteError(#[source] std::io::Error, PathBuf),

    #[error(
        "No tile sources found. Set sources by giving a database connection string on command line, env variable, or a config file."
    )]
    NoSources,
    #[error("Source path is not a file: {0}")]
    InvalidFilePath(PathBuf),

    #[error("Error {0} while parsing URL {1}")]
    InvalidSourceUrl(#[source] url::ParseError, String),

    #[error("Could not parse source path {0} as a URL")]
    PathNotConvertibleToUrl(PathBuf),

    #[error("Source {0} uses bad file {1}")]
    InvalidSourceFilePath(String, PathBuf),

    #[error("At least one 'origin' must be specified in the 'cors' configuration")]
    CorsNoOriginsConfigured,

    #[cfg(feature = "styles")]
    #[error("Walk directory error {0}: {1}")]
    DirectoryWalking(#[source] walkdir::Error, PathBuf),

    #[cfg(feature = "postgres")]
    #[error("The postgres pool_size must be greater than or equal to 1")]
    PostgresPoolSizeInvalid,

    #[cfg(feature = "postgres")]
    #[error("A postgres connection string must be provided")]
    PostgresConnectionStringMissing,

    #[cfg(feature = "postgres")]
    #[error("Failed to create postgres pool: {0}")]
    PostgresPoolCreationFailed(#[source] martin_core::tiles::postgres::PostgresError),

    #[cfg(feature = "fonts")]
    #[error("Failed to load fonts from {1}: {0}")]
    FontResolutionFailed(#[source] martin_core::fonts::FontError, PathBuf),

    #[cfg(feature = "pmtiles")]
    #[error("Failed to parse object store URL of {1}: {0}")]
    ObjectStoreUrlParsing(object_store::Error, String),
}
