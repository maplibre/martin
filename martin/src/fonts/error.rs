use std::path::PathBuf;

use pbf_font_tools::PbfFontError;

use super::CP_RANGE_SIZE;

#[derive(thiserror::Error, Debug)]
pub enum FontError {
    #[error("Font {0} not found")]
    FontNotFound(String),

    #[error("Font range start ({0}) must be <= end ({1})")]
    InvalidFontRangeStartEnd(u32, u32),

    #[error("Font range start ({0}) must be multiple of {CP_RANGE_SIZE} (e.g. 0, 256, 512, ...)")]
    InvalidFontRangeStart(u32),

    #[error(
        "Font range end ({0}) must be multiple of {CP_RANGE_SIZE} - 1 (e.g. 255, 511, 767, ...)"
    )]
    InvalidFontRangeEnd(u32),

    #[error(
        "Given font range {0}-{1} is invalid. It must be {CP_RANGE_SIZE} characters long (e.g. 0-255, 256-511, ...)"
    )]
    InvalidFontRange(u32, u32),

    #[error(transparent)]
    FreeType(#[from] pbf_font_tools::freetype::Error),

    #[error("IO error accessing {1}: {0}")]
    IoError(std::io::Error, PathBuf),

    #[error("Invalid font file {0}")]
    InvalidFontFilePath(PathBuf),

    #[error("No font files found in {0}")]
    NoFontFilesFound(PathBuf),

    #[error("Font {0} is missing a family name")]
    MissingFamilyName(PathBuf),

    #[error(transparent)]
    PbfFontError(#[from] PbfFontError),

    #[error(transparent)]
    ErrorSerializingProtobuf(#[from] pbf_font_tools::prost::DecodeError),
}
