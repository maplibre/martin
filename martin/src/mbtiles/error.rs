#[derive(thiserror::Error, Debug)]
pub enum MbtilesError {
    #[error(r"Unable to acquire connection to file: {0}")]
    AcquireConnError(String),

    #[error(transparent)]
    MbtilesLibraryError(#[from] mbtiles::MbtError),
}
