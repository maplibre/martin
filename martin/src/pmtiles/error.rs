use url::Url;

#[derive(thiserror::Error, Debug)]
pub enum PmtilesError {
    #[error(r"Error occurred in processing S3 source uri: {0}")]
    S3SourceError(String),

    #[cfg(feature = "pmtiles")]
    #[error(r"PMTiles error {0:?} processing {1}")]
    PmtilesLibraryError(pmtiles::PmtError, String),

    #[error(r"Unable to parse metadata in file {1}: {0}")]
    InvalidUrlMetadata(String, Url),

    #[error(r"Invalid tile coordinates")]
    InvalidTileCoordinates,
}
