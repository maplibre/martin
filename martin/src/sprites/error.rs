use std::path::PathBuf;

use spreet::SpreetError;
use spreet::resvg::usvg::Error as ResvgError;

#[derive(thiserror::Error, Debug)]
pub enum SpriteError {
    #[error("Sprite {0} not found")]
    SpriteNotFound(String),

    #[error("IO error {0}: {1}")]
    IoError(std::io::Error, PathBuf),

    #[error("Sprite path is not a file: {0}")]
    InvalidFilePath(PathBuf),

    #[error("Sprite {0} uses bad file {1}")]
    InvalidSpriteFilePath(String, PathBuf),

    #[error("No sprite files found in {0}")]
    NoSpriteFilesFound(PathBuf),

    #[error("Sprite {0} could not be loaded")]
    UnableToReadSprite(PathBuf),

    #[error("{0} in file {1}")]
    SpriteProcessingError(SpreetError, PathBuf),

    #[error("{0} in file {1}")]
    SpriteParsingError(ResvgError, PathBuf),

    #[error("Unable to generate spritesheet")]
    UnableToGenerateSpritesheet,

    #[error("Unable to create a sprite from file {0}")]
    SpriteInstError(PathBuf),
}
