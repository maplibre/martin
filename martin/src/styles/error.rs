use std::path::PathBuf;

#[derive(thiserror::Error, Debug)]
pub enum StyleError {
    #[error("Walk directory error {0}: {1}")]
    DirectoryWalking(walkdir::Error, PathBuf),
}
