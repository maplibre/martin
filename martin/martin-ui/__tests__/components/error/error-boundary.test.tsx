import { fireEvent, screen } from '@testing-library/react';
import { ErrorBoundary } from '@/components/error/error-boundary';
import { render } from '../../test-utils';

// Silence React error boundary logs for cleaner test output
let consoleErrorSpy: jest.SpyInstance;

beforeAll(() => {
  consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
});

afterAll(() => {
  consoleErrorSpy.mockRestore();
});

// Helper component to throw error
function ProblemChild({ shouldThrow }: { shouldThrow?: boolean }) {
  if (shouldThrow) {
    throw new Error('Test error thrown!');
  }
  return <div>All good</div>;
}

describe('ErrorBoundary', () => {
  it('renders children when no error occurs', () => {
    const { container } = render(
      <ErrorBoundary>
        <div>Safe content</div>
      </ErrorBoundary>,
    );
    expect(container).toMatchSnapshot();
    expect(screen.getByText('Safe content')).toBeInTheDocument();
  });

  it('catches errors and renders default fallback UI', () => {
    const { container } = render(
      <ErrorBoundary>
        <ProblemChild shouldThrow />
      </ErrorBoundary>,
    );
    expect(container).toMatchSnapshot();
    expect(screen.getByText('Something went wrong')).toBeInTheDocument();
    expect(
      screen.getByText(
        'An unexpected error occurred. Please try again or contact support if the problem persists.',
      ),
    ).toBeInTheDocument();
    expect(screen.getByText('Test error thrown!')).toBeInTheDocument();
    expect(screen.getByText('Try Again')).toBeInTheDocument();
  });

  it('calls onError prop when error is caught', () => {
    const onError = jest.fn();
    render(
      <ErrorBoundary onError={onError}>
        <ProblemChild shouldThrow />
      </ErrorBoundary>,
    );
    expect(onError).toHaveBeenCalledTimes(1);
    expect(onError.mock.calls[0][0]).toBeInstanceOf(Error);
    expect(onError.mock.calls[0][0].message).toBe('Test error thrown!');
    expect(onError.mock.calls[0][1]).toBeDefined();
  });

  it('renders custom fallback component if provided', () => {
    const CustomFallback = ({ error, retry }: { error?: Error; retry: () => void }) => (
      <div>
        <span>Custom fallback!</span>
        {error && <span>{error.message}</span>}
        <button onClick={retry} type="button">
          Retry now
        </button>
      </div>
    );
    const { container } = render(
      <ErrorBoundary fallback={CustomFallback}>
        <ProblemChild shouldThrow />
      </ErrorBoundary>,
    );
    expect(container).toMatchSnapshot();
    expect(screen.getByText('Custom fallback!')).toBeInTheDocument();
    expect(screen.getByText('Test error thrown!')).toBeInTheDocument();
    expect(screen.getByText('Retry now')).toBeInTheDocument();
  });

  it('resets error state and retries when retry button is clicked (default fallback)', () => {
    // Use a key to force remount after retry
    let key = 0;
    const { container, rerender } = render(
      <ErrorBoundary key={key}>
        <ProblemChild shouldThrow />
      </ErrorBoundary>,
    );
    // Should show error fallback
    expect(screen.getByText('Something went wrong')).toBeInTheDocument();

    // Click retry
    fireEvent.click(screen.getByText('Try Again'));

    // Rerender with no error and a new key to remount ErrorBoundary
    key++;
    rerender(
      <ErrorBoundary key={key}>
        <ProblemChild shouldThrow={false} />
      </ErrorBoundary>,
    );

    expect(container).toMatchSnapshot();
    expect(screen.queryByText('Something went wrong')).not.toBeInTheDocument();
    expect(screen.getByText('All good')).toBeInTheDocument();
  });

  it('resets error state and retries when retry button is clicked (custom fallback)', () => {
    const CustomFallback = ({ error, retry }: { error?: Error; retry: () => void }) => (
      <div>
        <span>Custom fallback!</span>
        {error && <span>{error.message}</span>}
        <button onClick={retry} type="button">
          Retry now
        </button>
      </div>
    );
    // Use a key to force remount after retry
    let key = 0;
    const { container, rerender } = render(
      <ErrorBoundary fallback={CustomFallback} key={key}>
        <ProblemChild shouldThrow />
      </ErrorBoundary>,
    );
    expect(screen.getByText('Custom fallback!')).toBeInTheDocument();

    // Click retry
    fireEvent.click(screen.getByText('Retry now'));

    // Rerender with no error and a new key to remount ErrorBoundary
    key++;
    rerender(
      <ErrorBoundary fallback={CustomFallback} key={key}>
        <ProblemChild shouldThrow={false} />
      </ErrorBoundary>,
    );

    expect(container).toMatchSnapshot();
    expect(screen.queryByText('Custom fallback!')).not.toBeInTheDocument();
    expect(screen.getByText('All good')).toBeInTheDocument();
  });
});
