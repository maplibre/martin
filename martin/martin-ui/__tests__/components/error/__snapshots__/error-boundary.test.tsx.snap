// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ErrorBoundary catches errors and renders default fallback UI 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,s,n,l,o,d,u,h)=&gt;{let m=document.documentElement,w=["light","dark"];function p(r){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&d?o.map(f=&gt;d[f]||f):o;k?(m.classList.remove(...S),m.classList.add(d&&d[r]?d[r]:r)):m.setAttribute(y,r)}),C(r)}function C(r){h&&w.includes(r)&&(m.style.colorScheme=r)}function a(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(l)p(l);else try{let r=localStorage.getItem(s)||n,y=u&&r==="system"?a():r;p(y)}catch(r){}})("class","theme","light",null,["light","dark"],null,true,true)
  </script>
  <div
    class="rounded-lg border bg-card text-card-foreground shadow-sm w-full max-w-md mx-auto mt-8"
  >
    <div
      class="flex flex-col space-y-1.5 p-6"
    >
      <div
        class="flex items-center space-x-2"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-triangle-alert h-5 w-5 text-destructive"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m21.73 18-8-14a2 2 0 0 0-3.48 0l-8 14A2 2 0 0 0 4 21h16a2 2 0 0 0 1.73-3"
          />
          <path
            d="M12 9v4"
          />
          <path
            d="M12 17h.01"
          />
        </svg>
        <div
          class="text-2xl font-semibold leading-none tracking-tight text-destructive"
        >
          Something went wrong
        </div>
      </div>
      <div
        class="text-sm text-muted-foreground"
      >
        An unexpected error occurred. Please try again or contact support if the problem persists.
      </div>
    </div>
    <div
      class="p-6 pt-0 space-y-4"
    >
      <div
        class="p-3 bg-destructive rounded-md"
      >
        <p
          class="text-sm text-destructive-foreground font-mono"
        >
          Test error thrown!
        </p>
      </div>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive h-9 px-4 py-2 has-[>svg]:px-3 bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 w-full"
        data-slot="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-refresh-cw h-4 w-4 mr-2"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 12a9 9 0 0 1 9-9 9.75 9.75 0 0 1 6.74 2.74L21 8"
          />
          <path
            d="M21 3v5h-5"
          />
          <path
            d="M21 12a9 9 0 0 1-9 9 9.75 9.75 0 0 1-6.74-2.74L3 16"
          />
          <path
            d="M8 16H3v5"
          />
        </svg>
        Try Again
      </button>
    </div>
  </div>
</div>
`;

exports[`ErrorBoundary renders children when no error occurs 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,s,n,l,o,d,u,h)=&gt;{let m=document.documentElement,w=["light","dark"];function p(r){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&d?o.map(f=&gt;d[f]||f):o;k?(m.classList.remove(...S),m.classList.add(d&&d[r]?d[r]:r)):m.setAttribute(y,r)}),C(r)}function C(r){h&&w.includes(r)&&(m.style.colorScheme=r)}function a(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(l)p(l);else try{let r=localStorage.getItem(s)||n,y=u&&r==="system"?a():r;p(y)}catch(r){}})("class","theme","light",null,["light","dark"],null,true,true)
  </script>
  <div>
    Safe content
  </div>
</div>
`;

exports[`ErrorBoundary renders custom fallback component if provided 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,s,n,l,o,d,u,h)=&gt;{let m=document.documentElement,w=["light","dark"];function p(r){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&d?o.map(f=&gt;d[f]||f):o;k?(m.classList.remove(...S),m.classList.add(d&&d[r]?d[r]:r)):m.setAttribute(y,r)}),C(r)}function C(r){h&&w.includes(r)&&(m.style.colorScheme=r)}function a(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(l)p(l);else try{let r=localStorage.getItem(s)||n,y=u&&r==="system"?a():r;p(y)}catch(r){}})("class","theme","light",null,["light","dark"],null,true,true)
  </script>
  <div>
    <span>
      Custom fallback!
    </span>
    <span>
      Test error thrown!
    </span>
    <button
      type="button"
    >
      Retry now
    </button>
  </div>
</div>
`;

exports[`ErrorBoundary resets error state and retries when retry button is clicked (custom fallback) 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,s,n,l,o,d,u,h)=&gt;{let m=document.documentElement,w=["light","dark"];function p(r){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&d?o.map(f=&gt;d[f]||f):o;k?(m.classList.remove(...S),m.classList.add(d&&d[r]?d[r]:r)):m.setAttribute(y,r)}),C(r)}function C(r){h&&w.includes(r)&&(m.style.colorScheme=r)}function a(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(l)p(l);else try{let r=localStorage.getItem(s)||n,y=u&&r==="system"?a():r;p(y)}catch(r){}})("class","theme","light",null,["light","dark"],null,true,true)
  </script>
  <div>
    All good
  </div>
</div>
`;

exports[`ErrorBoundary resets error state and retries when retry button is clicked (default fallback) 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,s,n,l,o,d,u,h)=&gt;{let m=document.documentElement,w=["light","dark"];function p(r){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&d?o.map(f=&gt;d[f]||f):o;k?(m.classList.remove(...S),m.classList.add(d&&d[r]?d[r]:r)):m.setAttribute(y,r)}),C(r)}function C(r){h&&w.includes(r)&&(m.style.colorScheme=r)}function a(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(l)p(l);else try{let r=localStorage.getItem(s)||n,y=u&&r==="system"?a():r;p(y)}catch(r){}})("class","theme","light",null,["light","dark"],null,true,true)
  </script>
  <div>
    All good
  </div>
</div>
`;
